### ISSUE TITLE
ActiveStorage: has_many_attached only persists the last added file when attaching via file i/o

### ISSUE THREAD

**Issue by KurtRMueller** [rails/rails#36806]
### Steps to reproduce

I've set up a sample repo here: https://github.com/KurtRMueller/active_storage_bug

Create a new rails 6.0.0.rc2 app. Install activestorage. Add the `has_many_attached` macro to a record. Go to the console, build a new record, start adding File/IO objects via the `attach` method, and then save the record. The first attachment will not be found on the disk, throwing a `ActiveStorage::FileNotFoundError (ActiveStorage::FileNotFoundError)`.

```ruby
% rails c
Running via Spring preloader in process 16465
Loading development environment (Rails 6.0.0.rc2)
2.6.3 :001 > u = User.new
   (2.4ms)  SELECT sqlite_version(*)
 => #<User id: nil, created_at: nil, updated_at: nil>
2.6.3 :002 > u.biographies.attach(io: File.open(Rails.root.join("test", "fixtures", "files", "biography1.txt")), filename: "biography1.txt")
 => #<ActiveStorage::Attached::Changes::CreateMany:0x00007f92b00469c0 @name="biographies", @record=#<User id: nil, created_at: nil, updated_at: nil>, @attachables=[{:io=>#<File:/Users/mueller.128/repos/asctech/active_storage_bug/test/fixtures/files/biography1.txt>, :filename=>"biography1.txt"}]>
2.6.3 :003 > u.biographies.attach(io: File.open(Rails.root.join("test", "fixtures", "files", "biography2.txt")), filename: "biography2.txt")
 => #<ActiveStorage::Attached::Changes::CreateMany:0x00007f92aec05038 @name="biographies", @record=#<User id: nil, created_at: nil, updated_at: nil>, @attachables=[#<ActiveStorage::Blob id: nil, key: nil, filename: "biography1.txt", content_type: "text/plain", metadata: {"identified"=>true}, byte_size: 10, checksum: "MJCeR48ILeEmUu+45nNyYA==", created_at: nil>, {:io=>#<File:/Users/mueller.128/repos/asctech/active_storage_bug/test/fixtures/files/biography2.txt>, :filename=>"biography2.txt"}]>
2.6.3 :004 > u.save
   (0.1ms)  begin transaction
  User Create (1.9ms)  INSERT INTO "users" ("created_at", "updated_at") VALUES (?, ?)  [["created_at", "2019-07-29 20:27:30.196911"], ["updated_at", "2019-07-29 20:27:30.196911"]]
  ActiveStorage::Attachment Load (0.6ms)  SELECT "active_storage_attachments".* FROM "active_storage_attachments" WHERE "active_storage_attachments"."record_id" = ? AND "active_storage_attachments"."record_type" = ? AND "active_storage_attachments"."name" = ?  [["record_id", 3], ["record_type", "User"], ["name", "biographies"]]
  ActiveStorage::Blob Create (0.7ms)  INSERT INTO "active_storage_blobs" ("key", "filename", "content_type", "metadata", "byte_size", "checksum", "created_at") VALUES (?, ?, ?, ?, ?, ?, ?)  [["key", "vcyrgwnh5psdfylo2thd4ca6wk0h"], ["filename", "biography1.txt"], ["content_type", "text/plain"], ["metadata", "{\"identified\":true}"], ["byte_size", 10], ["checksum", "MJCeR48ILeEmUu+45nNyYA=="], ["created_at", "2019-07-29 20:27:30.222822"]]
  ActiveStorage::Attachment Create (0.5ms)  INSERT INTO "active_storage_attachments" ("name", "record_type", "record_id", "blob_id", "created_at") VALUES (?, ?, ?, ?, ?)  [["name", "biographies"], ["record_type", "User"], ["record_id", 3], ["blob_id", 5], ["created_at", "2019-07-29 20:27:30.225392"]]
  ActiveStorage::Blob Create (0.1ms)  INSERT INTO "active_storage_blobs" ("key", "filename", "content_type", "metadata", "byte_size", "checksum", "created_at") VALUES (?, ?, ?, ?, ?, ?, ?)  [["key", "yb1cq10ztku7kybhz9i2oxupyofq"], ["filename", "biography2.txt"], ["content_type", "text/plain"], ["metadata", "{\"identified\":true}"], ["byte_size", 10], ["checksum", "4Xkl7dPxO3glrvsi+dEdEQ=="], ["created_at", "2019-07-29 20:27:30.226937"]]
  ActiveStorage::Attachment Create (0.1ms)  INSERT INTO "active_storage_attachments" ("name", "record_type", "record_id", "blob_id", "created_at") VALUES (?, ?, ?, ?, ?)  [["name", "biographies"], ["record_type", "User"], ["record_id", 3], ["blob_id", 6], ["created_at", "2019-07-29 20:27:30.228379"]]
  User Update (0.1ms)  UPDATE "users" SET "updated_at" = ? WHERE "users"."id" = ?  [["updated_at", "2019-07-29 20:27:30.228903"], ["id", 3]]
   (1.2ms)  commit transaction
  Disk Storage (1.6ms) Uploaded file to key: yb1cq10ztku7kybhz9i2oxupyofq (checksum: 4Xkl7dPxO3glrvsi+dEdEQ==)
Enqueued ActiveStorage::AnalyzeJob (Job ID: 97b82630-c081-4575-b580-89a5dbd20a9e) to Async(active_storage_analysis) with arguments: #<GlobalID:0x00007f92addf0e20 @uri=#<URI::GID gid://active-storage-bug/ActiveStorage::Blob/5>>
Enqueued ActiveStorage::AnalyzeJob (Job ID: 12800a7c-cb18-449d-bfdf-341804795c8a) to Async(active_storage_analysis) with arguments: #<GlobalID:0x00007f92acd62bf8 @uri=#<URI::GID gid://active-storage-bug/ActiveStorage::Blob/6>>
 => true
2.6.3 :005 >    (0.1ms)  SELECT sqlite_version(*)
   (0.1ms)  SELECT sqlite_version(*)
   (0.0ms)  SELECT sqlite_version(*)
  ActiveStorage::Blob Load (0.1ms)  SELECT "active_storage_blobs".* FROM "active_storage_blobs" WHERE "active_storage_blobs"."id" = ? LIMIT ?  [["id", 5], ["LIMIT", 1]]
Performing ActiveStorage::AnalyzeJob (Job ID: 97b82630-c081-4575-b580-89a5dbd20a9e) from Async(active_storage_analysis) enqueued at 2019-07-29T20:27:30Z with arguments: #<GlobalID:0x00007f92af06f1f0 @uri=#<URI::GID gid://active-storage-bug/ActiveStorage::Blob/5>>
  ActiveStorage::Blob Load (0.2ms)  SELECT "active_storage_blobs".* FROM "active_storage_blobs" WHERE "active_storage_blobs"."id" = ? LIMIT ?  [["id", 6], ["LIMIT", 1]]
Performing ActiveStorage::AnalyzeJob (Job ID: 12800a7c-cb18-449d-bfdf-341804795c8a) from Async(active_storage_analysis) enqueued at 2019-07-29T20:27:30Z with arguments: #<GlobalID:0x00007f92af09de88 @uri=#<URI::GID gid://active-storage-bug/ActiveStorage::Blob/6>>
   (0.1ms)  begin transaction
   (0.1ms)  begin transaction
  ActiveStorage::Blob Update (0.4ms)  UPDATE "active_storage_blobs" SET "metadata" = ? WHERE "active_storage_blobs"."id" = ?  [["metadata", "{\"identified\":true,\"analyzed\":true}"], ["id", 5]]
   (1.4ms)  commit transaction
  ActiveStorage::Blob Update (0.7ms)  UPDATE "active_storage_blobs" SET "metadata" = ? WHERE "active_storage_blobs"."id" = ?  [["metadata", "{\"identified\":true,\"analyzed\":true}"], ["id", 6]]
   (1.2ms)  commit transaction
Performed ActiveStorage::AnalyzeJob (Job ID: 97b82630-c081-4575-b580-89a5dbd20a9e) from Async(active_storage_analysis) in 16.83ms
Performed ActiveStorage::AnalyzeJob (Job ID: 12800a7c-cb18-449d-bfdf-341804795c8a) from Async(active_storage_analysis) in 15.72ms
u.biographies.first.download
  Disk Storage (0.4ms) Downloaded file from key: vcyrgwnh5psdfylo2thd4ca6wk0h
Traceback (most recent call last):
        1: from (irb):5
ActiveStorage::FileNotFoundError (ActiveStorage::FileNotFoundError)
2.6.3 :006 > u.biographies.last.download
  Disk Storage (6.1ms) Downloaded file from key: yb1cq10ztku7kybhz9i2oxupyofq
 => "biography2"
```

### Expected behavior

For any record with a `has_many_attached` macro, building a new record and then attaching File/IO objects via `attach` persists all attachments.

### Actual behavior

The first attachment is not persisted.

### System configuration
**Rails version**: 6.0.0.rc2

**Ruby version**: 2.6.3

**Comment by KurtRMueller** [comment]
I've had this happen to me with both a sqlite backend and a postgres backend as well.

**Comment by rwojnarowski** [comment]
Did you try to attach files as an arrays? Like so:
`u.biographies.attach([io: File.open(...)], [io: File.open(...)])`

I've got similar issue in my specs, i applied this as temp fix

**Comment by georgeclaghorn** [comment]
#36716 caused this.

**Comment by hidde-jan** [comment]
Since the current 5.2 behaviour will be replaced in 6.1, what is the recommended way of attaching new attachments without replacing the older ones?

**Comment by noma4i** [comment]
Any news here?

**Comment by akefeli** [comment]
I still have this issue in 6.0.3.2
```
irb(main):154:0> ActiveRecord::Base.transaction do
irb(main):155:1*  l.files.attach(io: File.open(p), filename: 'first.txt')
irb(main):156:1>  l.files.attach(io: File.open(p2), filename: 'second.txt')
irb(main):157:1> end
   (0.7ms)  BEGIN
  ActiveStorage::Blob Load (1.1ms)  SELECT `active_storage_blobs`.* FROM `active_storage_blobs` INNER JOIN `active_storage_attachments` ON `active_storage_blobs`.`id` = `active_storage_attachments`.`blob_id` WHERE `active_storage_attachments`.`record_id` = 28 AND `active_storage_attachments`.`record_type` = 'LearnableDocument' AND `active_storage_attachments`.`name` = 'files'
  ActiveStorage::Attachment Load (1.2ms)  SELECT `active_storage_attachments`.* FROM `active_storage_attachments` WHERE `active_storage_attachments`.`record_id` = 28 AND `active_storage_attachments`.`record_type` = 'LearnableDocument' AND `active_storage_attachments`.`name` = 'files'
  ActiveStorage::Blob Create (0.9ms)  INSERT INTO `active_storage_blobs` (`key`, `filename`, `content_type`, `metadata`, `byte_size`, `checksum`, `created_at`) VALUES ('bfksctzs6xkwgvl2y8g88alpipmf', 'first.txt', 'text/plain', '{\"identified\":true}', 7, 'I339G+V3uTntdla0c+EYlg==', '2020-07-13 22:14:48')
  ActiveStorage::Attachment Create (0.6ms)  INSERT INTO `active_storage_attachments` (`name`, `record_type`, `record_id`, `blob_id`, `created_at`) VALUES ('files', 'LearnableDocument', 28, 13591, '2020-07-13 22:14:48')
  ActiveStorage::Blob Create (0.5ms)  INSERT INTO `active_storage_blobs` (`key`, `filename`, `content_type`, `metadata`, `byte_size`, `checksum`, `created_at`) VALUES ('h37lv8ya83wk214842i2xn02inqw', 'second.txt', 'text/plain', '{\"identified\":true}', 10, '+bxkk3Mgva5MWmmYnj3/rg==', '2020-07-13 22:14:48')
  ActiveStorage::Attachment Create (0.5ms)  INSERT INTO `active_storage_attachments` (`name`, `record_type`, `record_id`, `blob_id`, `created_at`) VALUES ('files', 'LearnableDocument', 28, 13592, '2020-07-13 22:14:48')
  LearnableDocument Update (0.5ms)  UPDATE `learnable_documents` SET `learnable_documents`.`updated_at` = '2020-07-13 22:14:48' WHERE `learnable_documents`.`id` = 28
   (7.3ms)  COMMIT
  GCS Storage (421.6ms) Uploaded file to key: h37lv8ya83wk214842i2xn02inqw (checksum: +bxkk3Mgva5MWmmYnj3/rg==)
Enqueued ActiveStorage::AnalyzeJob (Job ID: 85a2dc3d-2322-4443-ae3a-e113ea45e862) to Resque(generic_fast) with arguments: #<GlobalID:0x000055b2caa19a28 @uri=#<URI::GID gid://cvl32x/ActiveStorage::Blob/13591>>
Enqueued ActiveStorage::AnalyzeJob (Job ID: e410a011-7cff-4f92-8d32-0ac28b25dd42) to Resque(generic_fast) with arguments: #<GlobalID:0x000055b2caa1c4a8 @uri=#<URI::GID gid://cvl32x/ActiveStorage::Blob/13592>>
=> true
irb(main):158:0> l.files.count
   (0.9ms)  SELECT COUNT(*) FROM `active_storage_attachments` WHERE `active_storage_attachments`.`record_id` = 28 AND `active_storage_attachments`.`record_type` = 'LearnableDocument' AND `active_storage_attachments`.`name` = 'files'
=> 2
irb(main):159:0> l.files.second.download
  GCS Storage (270.8ms) Downloaded file from key: h37lv8ya83wk214842i2xn02inqw
=> "SDADASDAS\n"
irb(main):160:0> l.files.first.download
  GCS Storage (133.5ms) Downloaded file from key: bfksctzs6xkwgvl2y8g88alpipmf
Traceback (most recent call last):
        1: from (irb):160
ActiveStorage::FileNotFoundError (ActiveStorage::FileNotFoundError)
```

**Comment by Eusebius1920** [comment]
This is still happening (Rails 6.0.3.2). The files are in the association, but if you try to access them, you get `ActiveStorage::FileNotFoundError` - even if attached with `#attach`. I guess it is not the same issue as #36716

**Comment by lenart** [comment]
You might be looking for `config.active_storage.replace_on_assign_to_many = false` configuration which is `true` by default in Rails 6. Setting this to `false` will persist old files when attaching new ones (without providing all the attachments explicitly).

**Comment by exocode** [comment]
...and how to check for duplicates in this case?

**Comment by gdotdesign** [comment]
I was having the same issue, and doing this fixed it for me (`6.1.3.2`):

> Did you try to attach files as an arrays? Like so: `u.biographies.attach([io: File.open(...)], [io: File.open(...)])`
> 
> I've got similar issue in my specs, i applied this as temp fix
